internal/resource/downloader/downloader.go

```diff
@@ -0,0 +1,158 @@
+// temporary file on closure.
+//
+// The resulting resource is verified to have the right hash and size.
+//
+// Returns [ErrUnexpectedHash] if the hash of the downloaded resource does not
+// match the expected hash.
+// Returns [ErrUnexpectedSize] if the size of the downloaded resource does not
+// match the expected size.
+func (d *ResourceDownloader) Download(
+	ctx context.Context,
+	url *url.URL,
+	sha384 string,
+	size int64,
+) (io.ReadCloser, error) {
+	tmpFile, err := os.CreateTemp("", "resource-")
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	defer func() {
+		// Always close the file, we no longer require to have it open for
+		// this purpose. Another process/method can take over the file.
+		tmpFile.Close()
+
+		// If the download was successful, we don't need to remove the file.
+		// It is the responsibility of the caller to remove the file.
+		if err == nil {
+			return
+		}
+
+		// Remove the temporary file if the download failed. If we can't
+		// remove the file, log a warning, so the operator can clean it up
+		// manually.
+		removeErr := os.Remove(tmpFile.Name())
+		if removeErr != nil {
+			d.logger.Warningf("failed to remove temporary file %q: %v", tmpFile.Name(), removeErr)
+		}
+	}()
+
+	d.logger.Debugf("downloading resource: %s", url)
+
+	// Force the sha256 and sha384 digest to be calculated on download.
+	digest, err := d.client.Download(ctx, url, tmpFile.Name())
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	d.logger.Debugf("downloaded resource: %s", url)
+
+	if digest.SHA384 != sha384 {
+		return nil, errors.Errorf(
+			"%w: %q, got %q", ErrUnexpectedHash, sha384, digest.SHA384,
+		)
+	}
+	if digest.Size != size {
+		return nil, errors.Errorf(
+			"%w: %q, got %q", ErrUnexpectedSize, size, digest.Size,
+		)
+	}
+
+	// Create a reader for the temporary file containing the resource.
+	tmpFileReader, err := newTmpFileReader(tmpFile.Name(), d.logger)
+	if err != nil {
+		return nil, errors.Errorf("opening downloaded resource: %w", err)
+	}
+
+	return tmpFileReader, nil
+}
+
+// DownloadResult contains information about a downloaded charm.
+type DownloadResult struct {
+	ReadCloser io.ReadCloser
+	SHA256     string
+	SHA384     string
+	Size       int64
+}
+
+// tmpFileReader wraps an *os.File and deletes it when closed.
+type tmpFileReader struct {
+	logger logger.Logger
+	*os.File
+}
+
+func newTmpFileReader(file string, logger logger.Logger) (*tmpFileReader, error) {
+	f, err := os.Open(file)
+	if err != nil {
+		return nil, err
+	}
+	return &tmpFileReader{
+		logger: logger,
+		File:   f,
+	}, nil
+}
+
+// Close closes the temporary file and removes it. If the file cannot be
+// removed, an error is logged.
+func (f *tmpFileReader) Close() (err error) {
+	defer func() {
+		removeErr := os.Remove(f.Name())
+		if err == nil {
+			err = removeErr
+		} else if removeErr != nil {
+			f.logger.Warningf("failed to remove temporary file %q: %v", f.Name(), removeErr)
+		}
```

Comment:
nit: just log the warning and move on. It's a temp file, so the OS will clean this up on restart or periodically (if enabled).

I'm happy to leave this as is though.

Reply:
Nice, good to know for the future, will leave it as it is here.